src/dlgIRC.cpp

```diff
@@ -61,265 +90,455 @@ dlgIRC::dlgIRC()
+    commandParser->addCommand(IrcCommand::Trace, "TRACE (<target>)");
+    commandParser->addCommand(IrcCommand::Users, "USERS (<server>)");
+    commandParser->addCommand(IrcCommand::Version, "VERSION (<user>)");
+    commandParser->addCommand(IrcCommand::Who, "WHO <mask>");
+    commandParser->addCommand(IrcCommand::Whois, "WHOIS <user>");
+    commandParser->addCommand(IrcCommand::Whowas, "WHOWAS <user>");
+
+    commandParser->addCommand(IrcCommand::Custom, "MSG <target> <message...>");  // replaces the old /msg command.
+    commandParser->addCommand(IrcCommand::Custom, "CLEAR (<buffer>)");  // clears the given buffer, or the current active if none are given.
+    commandParser->addCommand(IrcCommand::Custom, "CLOSE (<buffer>)");  // closes the buffer and removes it from the list, uses current active buffer if none are given.
+    commandParser->addCommand(IrcCommand::Custom, "RECONNECT");  // Issues a Quit command and closes the IRC connection then reconnects to the IRC server.
+    commandParser->addCommand(IrcCommand::Custom, "HELP (<command>)");  // displays some help information about a given command or lists all available commands.
 }
 
-void dlgIRC::onConnected()
+void dlgIRC::setupBuffers()
 {
-    connection->sendCommand(IrcCommand::createJoin("#mudlet"));
+    bufferModel = new IrcBufferModel(connection);
+    connect(bufferModel, SIGNAL(added(IrcBuffer*)), this, SLOT(slot_onBufferAdded(IrcBuffer*)));
+    connect(bufferModel, SIGNAL(removed(IrcBuffer*)), this, SLOT(slot_onBufferRemoved(IrcBuffer*)));
+    bufferList->setModel(bufferModel);
+    // keep the command parser aware of the context
+    connect(bufferModel, SIGNAL(channelsChanged(QStringList)), commandParser, SLOT(setChannels(QStringList)));
+    // keep track of the current buffer, see also onBufferActivated()
+    connect(bufferList->selectionModel(), SIGNAL(currentChanged(QModelIndex,QModelIndex)), this, SLOT(slot_onBufferActivated(QModelIndex)));
+    // create a server buffer for non-targeted messages...
+    IrcBuffer* serverBuffer = bufferModel->add( connection->host() );
+    connect(bufferModel, SIGNAL(messageIgnored(IrcMessage*)), serverBuffer, SLOT(receiveMessage(IrcMessage*)));
 }
 
-void dlgIRC::onMessageReceived(IrcMessage* msg)
+bool dlgIRC::processCustomCommand(IrcCommand* cmd)
 {
-    switch (msg->type()) {
-    case IrcMessage::Type::Join: {
-        IrcJoinMessage* rmsg = static_cast<IrcJoinMessage*>(msg);
-        slot_joined(rmsg->nick(), rmsg->channel());
-        break;
+    if (cmd->type() != IrcCommand::Custom || cmd->parameters().count() < 1) {
+        return true;
     }
-    case IrcMessage::Type::Notice: {
-        IrcNoticeMessage* rmsg = static_cast<IrcNoticeMessage*>(msg);
-        irc_gotMsg(rmsg->nick(), rmsg->target(), rmsg->content());
-        break;
+
+    const QString cmdName = QString(cmd->parameters().at(0)).toUpper();
+    if (cmdName == "CLEAR") {
```

Comment:
Actually was because QString is illegal in `switch` statements.   
The only way to know what command you're actually handling when dealing with the `IrcCommand::Custom` type is to inspect the command string itself for the base command name.